### ISSUE TITLE
dnsdist crash with used readn2 on non-blocking socket, got EAGAIN

### ISSUE THREAD

**Issue by rygl** [PowerDNS/pdns#3560]
Hi all,
I have experienced a crash of dnsdist today:

`Mar 13 07:06:17 rzt-lvs1 dnsdist[25028]: terminate called after throwing an instance of 'std::runtime_error'`
`Mar 13 07:06:17 rzt-lvs1 dnsdist[25028]: what():  used readn2 on non-blocking socket, got EAGAIN`
`Mar 13 07:06:17 rzt-lvs1 systemd[1]: dnsdist.service: main process exited, code=killed, status=6/ABRT`

version: 0.0.790g37bf069-1pdns.jessie

The gdb output  is downloadable here: (thread apply all bt full)
http://owncloud.rygl.net/owncloud/index.php/s/lI0MZvD6FzdIXiV

I am sorry it it is something what has been already fixed in newer builds but I have not seen this error yet.

Cheers
Ales

**Comment by rgacogne** [comment]
Hi Aleš,

Thank you for reporting this, I am looking into it!

**Comment by rygl** [comment]
Hi Remi,
what was really bad this time was that even after dnsdist restart by systemd it was not working correctly and I had to restart it manually...
I have also a packet capture from this moment.
... and shortly before posting this it crashed again throwing the same error.
A.

**Comment by rgacogne** [comment]
I don't get it for now, we have very few calls to readn2() in dnsdist and I don't see how any of them could be on an non-blocking fd. What kernel are you using? Did you upgrade dnsdist recently, and if so from which version?

**Comment by rygl** [comment]
Hi Remi,
The kernel is latest distribution kernel from Debian Jessie: 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u4

I downgraded from  0.0.816g0048c4c after the Friday's issue with filedescriptors.


I have one more core from today, I will provide you with the gdb output. Maybe you will find something there ...

**Comment by rygl** [comment]
Promised gdb output:
http://owncloud.rygl.net/owncloud/index.php/s/chsAZHoitWuh6FO

**Comment by ahupowerdns** [comment]
the readn2 is on the pipe where tcpclientthreads retrieve work that needs to be done. We have >700 threads running at that point. The crash is in an abort that looks related to an exception escaping a thread, but hard to tell. Most obvious question is how we end up running 700 tcp client threads. I spot at least 4092 file descriptors in use which might mean that we hit the magic "4096" on attempting to spawn a new thread.

**Comment by ahupowerdns** [comment]
ok didn't read the beginning of the ticket - the pipe we use to distribute queries to tcp threads is set to nonblocking (why?) and the reading thread is using readn2 which assumes a blocking socket. Now it may make sense to make the *write* part of that pipe nonblocking, but not the reading part. But @rgacogne will know what to do!

**Comment by rygl** [comment]
Hi,
there is something happening about TCP in our network. It started on Friday when I run out of filedescriptors (>4096) without any apparent reason. I am running dnsdist in production for couple of moths and never had problems with that.  So I have increased FDs to 65k and also setMaxTCPClientThreads() to 2048. Let me know if these values are ok from yout pouint of view.

I made a trace when having this FD issue and I am afraid that this is the first time when the pseudo-random subdomain attack appears on TCP! Have you ever seen this? 

Feel free to download the sample here: 
http://owncloud.rygl.net/owncloud/index.php/s/U9An5nK1H1ZRJ5l

Cheers
A.

**Comment by rgacogne** [comment]
> the pipe we use to distribute queries to tcp threads is set to nonblocking (why?) and the reading thread is using readn2 which assumes a blocking socket. Now it may make sense to make the write part of that pipe nonblocking, but not the reading part.

Yes, the writing FD is set to non-blocking because we can't afford to block on a full pipe, but the reading FD is not, and we only use readn2() on the reading part, so I don't see how we get EAGAIN on this FD for now.

> Let me know if these values are ok from yout pouint of view.

Yes, they seem fine, you should not reach the FD limit with these. If you do, we are probably doing something wrong.

> I made a trace when having this FD issue and I am afraid that this is the first time when the pseudo-random subdomain attack appears on TCP! Have you ever seen this?

Looking at the PCAP, it looks like most of the queries are from the same source IP, so it might be someone using a tunnel over DNS or something like that?

**Comment by rgacogne** [comment]
I've added a bit of information in 18861f97234a05547611b7e3f3eecef867a40e1b, if you have the opportunity to upgrade we might get a better understanding of what's going on.

**Comment by rygl** [comment]
> Looking at the PCAP, it looks like most of the queries are from the same source IP, so it might be someone using a tunnel over DNS or something like that?
> 

I not not think it is a tunnel. It would look different, the queries here are pointing to the same domain and they are random and small. In case of a tunnel it would be a TXT query. It is an open resolver misused on TCP which scares me. I am already pushing the customer to close it immediately. 

I will upgrade to your improved version asap and come back.
A.

**Comment by rgacogne** [comment]
I have investigated this issue a bit more, using additional info kindly provided by @rygl.
With the information added in 18861f97234a05547611b7e3f3eecef867a40e1b, we can now confirm that the error occurs because the reading side of the TCP acceptor pipe has been somehow made non-blocking:
```
Error reading from TCP acceptor pipe (7286) in non-blocking mode: used readn2 on non-blocking socket, got EAGAIN
```
Looking at the memory dump corresponding to this error, I see that the corresponding file descriptor, 7286, is also present in the `sockets` map of another TCP thread, as a TCP socket to a backend. Hence why it's now non-blocking.

My current theory is that we open a TCP socket to a backend, close it but fail to remove the descriptor from the `sockets` map because an exception is raised in `setupTCPDownstream()`, leaving a dangling FD in the map.
The same descriptor number is assigned as the reading side of a pipe for a new TCP thread.
Meanwhile the existing thread tries to use the dangling socket, fails to write to it because it is now only usable for read operations and therefore closes it. A new call to `setupTCPDownstream()` assigns this same descriptor number to a TCP connection to a backend, and makes it non-blocking.
The next time the other TCP thread calls `readn2()`, it fails with this error as we are asking for a blocking read on a non-blocking descriptor.

I just opened #3634 with a fix for that case.
